src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxy.java

```diff
@@ -333,19 +354,73 @@ private ListShardsResult listShards(final String nextToken) {
      */
     @Override
     public Shard getShard(String shardId) {
-        if (this.listOfShardsSinceLastGet.get() == null) {
-            //Update this.listOfShardsSinceLastGet as needed.
-            this.getShardList();
+        if (this.cachedShardMap == null) {
+            synchronized (this) {
+                if (this.cachedShardMap == null) {
+                    this.getShardList();
+                }
+            }
+        }
+
+        Shard shard = cachedShardMap.get(shardId);
+        if (shard == null) {
+            if (cacheMisses.incrementAndGet() > MAX_CACHE_MISSES_BEFORE_RELOAD || cacheNeedsTimeUpdate()) {
+                synchronized (this) {
+                    shard = cachedShardMap.get(shardId);
+
+                    //
+                    // If after synchronizing we resolve the shard, it means someone else already got it for us.
+                    //
+                    if (shard == null) {
+                        LOG.info("To many shard map cache misses or cache is out of date -- forcing a refresh");
+                        this.getShardList();
```

Comment:
Same as above can do without this.

Reply:
It's he same thing as above.  If another processor got in first, don't redo their work.